relayer/src/supervisor.rs

```diff
@@ -195,70 +205,128 @@ impl Supervisor {
+            let channels = chain.query_channels(req.clone())?;
             for channel in channels {
-                let client = channel_connection_client(
-                    &channel.port_id,
-                    &channel.channel_id,
-                    chains.a.as_ref(),
-                )?
-                .client;
-                if client.client_state.chain_id() != chains.b.id() {
-                    continue;
-                }
+                self.spawn_workers_for_channel(chain.clone(), channel)?;
+            }
+        }
+
+        Ok(())
+    }
 
-                // create the client object and spawn worker
-                let client_object = Object::Client(Client {
-                    dst_client_id: client.client_id.clone(),
-                    dst_chain_id: chains.a.id(),
-                    src_chain_id: client.client_state.chain_id(),
-                });
-                let worker = Worker::spawn(chains.clone(), client_object.clone());
-                self.workers.entry(client_object).or_insert(worker);
-
-                // create the path object and spawn worker
-                let path_object = Object::UnidirectionalChannelPath(UnidirectionalChannelPath {
-                    dst_chain_id: chains.b.id(),
-                    src_chain_id: chains.a.id(),
-                    src_channel_id: channel.channel_id.clone(),
-                    src_port_id: channel.port_id.clone(),
-                });
-                let worker = Worker::spawn(chains.clone(), path_object.clone());
-                self.workers.entry(path_object).or_insert(worker);
+    /// Spawns all the [`Worker`]s that will handle a given channel for a given source chain.
+    fn spawn_workers_for_channel(
+        &mut self,
+        chain: Box<dyn ChainHandle>,
+        channel: IdentifiedChannelEnd,
+    ) -> Result<(), BoxError> {
+        trace!(
+            "Fetching connection_client for channel {:?} of chain {}",
+            channel,
+            chain.id()
+        );
+
+        let client_res =
+            channel_connection_client(&channel.port_id, &channel.channel_id, chain.as_ref());
+
+        let client = match client_res {
+            Ok(conn_client) => conn_client.client,
+            Err(Error::ConnectionNotOpen(..)) | Err(Error::ChannelNotOpen(..)) => {
+                // These errors are silent.
+                // Simply ignore the channel and return without spawning the workers.
+                warn!(
+                    "ignoring channel {} because it is not open (or its connection is not open)",
+                    channel.channel_id
+                );
+
+                return Ok(());
+            }
+            Err(e) => {
+                // Propagate errors.
+                return Err(format!(
+                    "unable to spawn workers for channel/chain pair '{}'/'{}' due to error: {:?}",
+                    channel.channel_id,
+                    chain.id(),
+                    e
+                )
```

Comment:
I think we should `error!` here or in the caller and continue best effort to create workers and relay with those. The way it is the supervisor exits.

Reply:
Yeah we should handle this error in the caller.

Reply:
The only possible error that we propagate here is currently `Error::QueryFailed`. This can have one of three causes:

1. `query_channel` failed, or
2. `query_connection` failed, or
3. `query_client_state` failed.

I think all these failure cases should be handled with a retry inside method `channel_connection_client` (or deeper in the stack). And `channel_connection_client` should return `Error::QueryFailed` only if the number of retries was exhausted.

Until we have the retry-based fix (https://github.com/informalsystems/ibc-rs/issues/712), I've added code to handle the propagated error in the caller -- 2373a61a. Let me know if this is good. It should be safe for the supervisor to continue (because returning an error to the supervisor/caller means no single worker was spawned for the given channel, so there won't be some partially incomplete state left behind).

Reply:
looks good. in the future, even with retries, we probably don't want the supervisor to exit on the failures ^